### ISSUE TITLE
Add n_samples argument for all perturbation based methods like it exists in ShapleyValueSampling

### ISSUE THREAD

**Issue by hal-314** [pytorch/captum#505]
## 🚀 Feature

Add `n_samples` argument for all perturbation based methods like it exists in ShapleyValueSampling. 

## Motivation

Perturbation based algorithms compute feature attribution by perturbing input features. So, depending in how is perturbed, feature attribution varies. To make a more robust estimation, perturbations should happen several times. For example, [scikit-learn feature permutation function](https://scikit-learn.org/stable/modules/generated/sklearn.inspection.permutation_importance.html) repeats permutation 5 times by default.

Finally, it'd be consistent with ShapleyValueSampling.
## Pitch

Implement `n_samples` argument for other perturbation algorithm in addition than ShapleyValueSampling.

## Alternatives

Do it manually by subclassing every algorithm (`FeaturePermutation`, `FeatureAblation` and `Occlusion`) and overriding `attribute` method. Then, the new attribution method would call several times the base implementations and average the results.

**Comment by bilalsal** [comment]
Thanks @hal-314 for the good suggestion!

Captum provides a generic mechanism to achieve the desired behavior, called [NoiseTunnel](https://captum.ai/docs/algorithms#noise-tunnel)
The [NoiseTunnel.attribute() method](https://captum.ai/api/noise_tunnel.html) takes `n_samples` as a parameter.

Hope this helps

**Comment by hal-314** [comment]
Thanks for the tip @bilalsal! I knew about `NoiseTunnel` but I didn't realize that it could be use to achieve this!

From user perspective, using `NoiseTunnel` with `FeaturePermutation` or any other perturbation algorithm to achieve `n_samples` behavior may not be obvious when `ShapleyValueSampling` has `n_samples`. It would be nice to put a notice in the docs. I discarded `NoiseTunnel` as seems to be focus on adding noise to samples.

I leave this issue open as I think that current situation isn't user friendly. Feel free to close this issue.

**Comment by vivekmig** [comment]
Hi @hal-314 , just wanted to get more context on this issue. For FeatureAblation and Occlusion, the perturbations don't involve any randomness and should be the same when repeating attribution for the same input and baselines. As long as the model forward pass is deterministic, we should expect multiple samples to result in the same attribution results. Given this, I'm not sure if it would make sense to add n_samples for these methods directly. What was the use-case that you had in mind for these? For FeaturePermutation, I think this definitely makes sense and is something we could look into. As @bilalsal mentioned, NoiseTunnel could potentially serve as a temporary workaround, but there's a couple of issues that might affect this approach. Particularly, since NoiseTunnel expands the input batch and provides this to the attribution method

**Comment by vivekmig** [comment]
Hi @hal-314 , just wanted to get more context on this issue. For FeatureAblation and Occlusion, the perturbations don't involve any randomness and should be the same when repeating attribution for the same input and baselines. As long as the model forward pass is deterministic, we should expect multiple samples to result in the same attribution results. What was the use-case that you had in mind for n_samples in these methods? 

For FeaturePermutation, I think this definitely makes sense and is something we could look into. As @bilalsal mentioned, NoiseTunnel could potentially serve as a workaround. A couple of things to keep in mind with this approach, NoiseTunnel expands the input internally and would essentially provide FeaturePermutation with the expanded original batch repeated n_samples times. This could constrain n_samples by memory limitations to fit the full expanded batch in memory until this issue #497 is addressed. Also, if FeaturePermutation is used with a forward function that returns a scalar value per batch (e.g. loss), the result would now correspond to applying a single permutation to the expanded batch, rather than averaging over multiple independent permutations on the original batch. In many cases, these two approaches will be similar, but there are cases where these could be different.

**Comment by hal-314** [comment]
@vishwakftw Rereading again FeatureAblation and Occlusion, I think you are right and n_samples argument doesn't make sense. I mainly used FeaturePermutation. As FeatureAblation and Occlusion were part of the same family (perturbation algorithms), I jumped to the conclusions and assume that they'd benefit from n_samples. Sorry for the misunderstanding. 

> For FeaturePermutation, I think this definitely makes sense and is something we could look into.

That's nice!

**Comment by vishwakftw** [comment]
@hal-314, you seem to have tagged the wrong person. :-)

**Comment by NarineK** [comment]
@hal-314, has this issue already been solved ? If so, can we, please, close this issue.
Thank you :)

**Comment by hal-314** [comment]
@NarineK FeaturePermutation` still doesn't have `n_samples` like ShapleyValueSampling. See the [api](https://captum.ai/api/feature_permutation.html). Feel free to close the issue if you think that isn't important.

**Comment by craymichael** [comment]
Apologies for the long delay here. A PR is now open at #1866 to add `n_samples` directly to `FeaturePermutation.attribute`, with the default staying at 1 so existing behavior remains unchanged.

The implementation averages multiple independent permutation-attribution estimates, so it matches the intent of repeatedly sampling permutations from the batch empirical feature distribution without requiring users to call `attribute` in a loop. I’ll leave this issue open for the PR to close when it lands.
